### PR TITLE
Added SPN authentication & relevant in-depth documentation.

### PR DESCRIPTION
**Pull request by Virepri**
I would've loved to add the ability to yank the certificate from the machine's certpool, but currently, this is broken on Windows.

### CODE REVIEW

File: cmd/login.go [hunk 1]
```diff
@@ -57,13 +57,22 @@ func init() {
 	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.aadEndpoint, "aad-endpoint", "", "the Azure active directory endpoint to use for OAuth user interactive login")
 	// Use identity which aligns to Azure powershell and CLI.
 	lgCmd.PersistentFlags().BoolVar(&loginCmdArgs.identity, "identity", false, "log in using virtual machine's identity, also known as managed service identity (MSI)")
+	// Use SPN certificate to log in.
+	lgCmd.PersistentFlags().BoolVar(&loginCmdArgs.certificate, "certificate", false, "log in using a certificate as a form of Service Principal Name (SPN) auth")
```
Comment (zezha-msft) [inline cmd/login.go]: These flags seem a bit redundant, right? We can infer that we are trying to use a certificate with the existed of cert-path and cert-pass

File: cmd/login.go [hunk 2]
```diff
@@ -57,13 +57,22 @@ func init() {
...
+	lgCmd.PersistentFlags().BoolVar(&loginCmdArgs.secret, "secret", false, "log in using a client secret as a form of Service Principal Name (SPN) auth")
```
Comment (zezha-msft) [inline cmd/login.go]: Same for this one, if a client-secret is provided, we can infer that the user wants to use secret to login.

File: README.md [hunk 3]
```diff
@@ -180,6 +180,51 @@ export https_proxy=<user>:<pass>@<proxy IP>:<proxy port>
...
+    4b. Create the certificate
+    
+    openssl req -new -key "example-app.key" -subj "/CN=example-app" -out "example-app.csr"
+    openssl x509 -req -in "example-app.csr" -signkey "example-app.key" -out "example-app.crt" -days 10000
+    
+    4c. Create the PKCS12 version of the certificate containing the private key
+    
+    openssl pkcs12 -export -out "example-app.pfx" -inkey "example-app.key" -in "example-app.crt" -passout <password>
+    
+**4**. Register the app as a contributor to any [resource groups or storage accounts](https://ms.portal.azure.com/#blade/HubsExtension/BrowseResourceGroupBlade/resourceType/Microsoft.Resources%2Fsubscriptions%2FresourceGroups) you plan to use with AzCopy.
```
Comment (zezha-msft) [inline README.md]: @jiacfan I'm not totally sure Contributor is the right role, shouldn't it be "Blob Data Contributor"?

File: README.md [hunk 4]
```diff
@@ -180,6 +180,51 @@ export https_proxy=<user>:<pass>@<proxy IP>:<proxy port>
...
+**2**. Ensure the app registration is labeled as a Web App/API. The sign-on URL does not matter, simply set it to http://whatever
+
+![Example registration](images/registration.png)
+
+**3**. Add public keys or client secrets under the "keys" section of the application.
+
+![Keys](images/keys.png)
+
+To generate a private & public keypair for AzCopy, perform the following steps in a command line, replacing example-app with your app name:
+
```
Comment (zezha-msft) [inline README.md]: Would it be simpler to refer to [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest#certificate-based-authentication) instead?

File: README.md [hunk 5]
```diff
@@ -180,6 +180,51 @@ export https_proxy=<user>:<pass>@<proxy IP>:<proxy port>
...
+Client secrets:
+
+```
+./azcopy login --secret --application-id <your application ID> --client-secret <your secret>
+```
+
+Certificates:
+
+```
+./azcopy login --certificate --application-id <your application ID> --certificate-path <your certificate's path, can be relative> --certificate-password <your certificate's password, not a mandatory parameter>
```
Comment (zezha-msft) [inline README.md]: rename --secret and --certificate to --service-principle

File: cmd/login.go (hunk 2 above)
Comment (jiacfan) [inline cmd/login.go]: Could this be --service-principal as Azure CLI [https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest](https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest);?

File: cmd/login.go [hunk 6]
```diff
@@ -57,13 +57,22 @@ func init() {
...
 	// Object ID of user-assigned identity.
 	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.identityObjectID, "identity-object-id", "", "object ID of user-assigned identity")
 	// Resource ID of user-assigned identity.
 	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.identityResourceID, "identity-resource-id", "", "resource ID of user-assigned identity")
 
+	//login with SPN
+	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.applicationID, "application-id", "", "application ID of user-assigned identity")
+	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.certPath, "certificate-path", "", "path to certificate for SPN authentication")
+	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.certPass, "certificate-password", "", "certificate password for SPN authentication")
+	lgCmd.PersistentFlags().StringVar(&loginCmdArgs.clientSecret, "client-secret", "", "client secret for SPN authentication")
```
Comment (jiacfan) [inline cmd/login.go]: @Ze please review this with secret team, and expire logic might be involved in AzCopy to protect the secret.

File: README.md (hunk 3 above)
Comment (Virepri) [reply README.md]: Contributor in my testing was the only one that would allow successful authentication.

File: cmd/login.go (hunk 1 above)
Comment (Virepri) [reply cmd/login.go]: Resolved in [@1c8b21](https://github.com/Azure/azure-storage-azcopy/pull/332/commits/1c8b217cd60befb694f9a063a5603dca8258958f), flags merged

File: cmd/login.go (hunk 2 above)
Comment (Virepri) [reply cmd/login.go]: Resolved in [@1c8b21](https://github.com/Azure/azure-storage-azcopy/pull/332/commits/1c8b217cd60befb694f9a063a5603dca8258958f), flags merged

File: README.md (hunk 4 above)
Comment (Virepri) [reply README.md]: Azure CLI is now referred to within the readme, while maintaining the alternative setup.

**Comment by adreed-msft** [comment]
Closing due to duplicate in #383 from work account.

**Pull request closed without merging**